sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala

```diff
@@ -268,20 +281,24 @@ class SortBasedAggregator(
         if (hasNextInput || hasNextAggBuffer) {
           // Find smaller key of the initialAggBufferIterator and initialAggBufferIterator
           groupingKey = findGroupingKey()
-          result = new AggregationBufferEntry(groupingKey, makeEmptyAggregationBuffer)
+          initializeBuffer(sortBasedAggregationBuffer)
+          result = new AggregationBufferEntry(groupingKey, sortBasedAggregationBuffer)
 
           // Firstly, update the aggregation buffer with input rows.
-          while (hasNextInput &&
-            groupingKeyOrdering.compare(inputIterator.getKey, groupingKey) == 0) {
-            processRow(result.aggregationBuffer, inputIterator.getValue)
-            hasNextInput = inputIterator.next()
+          if (aggregateMode != 1) {
+            var findNextPartition = false
+            while (hasNextInput && !findNextPartition) {
+              processRow(result.aggregationBuffer, inputIterator.getValue)
+              hasNextInput = inputIterator.next()
+              findNextPartition = inputIterator.getKey != groupingKey
+            }
```

Comment:
Hmm, the change of using `aggregateMode` is related to performance?

Reply:
I feel that `aggregateMode` makes the code harder to read.

Reply:
yea, it is used to avoid unnecessary grouping key comparation. Before, it gets grouping key by comparing input rows and aggregation buffer, but it do compare again when updating or merging aggregation buffer.